### ISSUE TITLE
Support npe_none

### ISSUE THREAD

**Issue by fwilliams** [fwilliams/numpyeigen#36]
Support `npe_none` type alonside `dense_*` and `sparse_*` allowing the user to have default null arguments.

Need to also set the type of the input to `none` and allow checking (`npe::is_none(v)`). 
This is sort of weird with `npe_matches` since we'll also enforce `none`. Maybe we want to not do this and have sematics which match non None types and disregard None types.

**Comment by fwilliams** [comment]
Valid new syntax:
* `npe_arg(name, [dense|sparse]_f32, [dense|sparse]_f64, none)`
* `npe_default_arg(name, [dense|sparse]_f32, [dense|sparse]_f64, none, npe::none())`
* `npe_[default_]arg(name, npe_matches_or_none(other))` is okay since you are asking to either match the other type or be none

New Error Case:
* `npe_[default_]arg(name, npe_matches(other_arg))` if `other` arg can be `none` is disallowed since in that case, we don't have a type to match to

**Comment by fwilliams** [comment]
I've changed my mind on this issue. Introducing a NULL value in a statically typed system is problematic. If an argument is either a numpy `array` or `None`, then we need a sane way to translate it into C++. The two ways I can think of doing this are: 
1. Introduce a special NULL Eigen matrix type and use that in the None branches.
2. Introduce a flag set at runtime which indicates if an input is NULL and pass in an arbitrary Eigen matrix.

Case (1) can cause tricky compiler errors if the user writes general code since there will exist a version of that code which gets called with the NULL Type. For example, if a user calles a function which accepts a `MatrixBase` of a specific range of types, the NULL Matrix may satisfy certain compile time assumptions and fail to compile. In the worst case, the NULL type may violate runtime preconditions and lead to hard-to-find bugs.

Case (2) let's a user write general code, but if they call code with a None input, this would be undefined behavior since the user cannot make assumptions about the matrix being passed in.

Personally, I'm unhappy with both these solutions. After thinking about the issue some more, I realized the reason we want a None type is to signal to the callee that the input is to be ignored. Another way of doing this is just to pass in an empty matrix. Currently this can be done with 
```
npe_default_arg(<name>, <arg1>, ..., <argN>, <default_value>)
```
At the moment, it's cumbersome to write the expression for `<default_value>`. e.g. `pybind11::array(pybind11::dtype("float32"), std::array<int, 2>(0, 0), std::array<int, 2>(0, 0))`. Instead, I plan on making simplified ways to initialize empty arrays: .e.g. `npe::dense_array(npe::float32, npe::row_major, 0, 0)` and to support `npe_default_arg(a, npe_empty_like(b))` which says the default value is an empty matrix like `b`.
